Add App Dev hosting for 4.x (#834)
medic/cht-docs: content/en/apps/guides/hosting/4.x/app-developer.md

---
title: "App Developer Hosting in CHT 4.x"
linkTitle: "App Developer Hosting"
weight: 40
aliases:
- /apps/guides/hosting/app-developer
description: >
  Hosting the CHT when developing apps
---

> {{% alert title="Note" %}} This guide assumes you are a CHT app developer wanting to either run concurrent instances of the CHT, or easily be able to switch between different instances without loosing any data while doing so. To do development on the CHT core itself, see the [development guide]({{< relref "apps/guides/hosting/core-developer.md" >}}).

Reviewer: ```suggestion
{{% alert title="Note" %}} This guide assumes you are a CHT app developer wanting to either run concurrent instances of the CHT, or easily be able to switch between different instances without losing any data while doing so. To do development on the CHT Core Framework itself, see the [development guide]({{< relref "apps/guides/hosting/core-developer.md" >}}).
```

To deploy the CHT 3.x in production, see either [AWS hosting]({{< relref "apps/guides/hosting/3.x/self-hosting.md" >}}) or [Self hosting]({{< relref "apps/guides/hosting/3.x/ec2-setup-guide.md" >}}). 4.x production hosting guides are coming soon!{{% /alert %}}

## Getting started

> Be sure to meet the [CHT hosting requirements]({{< relref "apps/guides/hosting/requirements" >}}) first. As well, if any other CHT 4.x instances using Docker Compose are running locally, stop them otherwise port, storage volume and container name conflicts may occur. To stop ALL containers, you can use `docker kill $(docker ps -q)`.

Reviewer: ```suggestion
Be sure to meet the [CHT hosting requirements]({{< relref "apps/guides/hosting/requirements" >}}) first. To avoid conflicts, ensure that all other CHT 4.x instances are stopped. To stop ALL containers, you can use

```shell
docker kill $(docker ps -q)
```

```

Reviewer: That suggestion didn't get formatted correctly. Suggest new line for the code so it doesn't get affected by a line break.

After meeting these requirements, create a directory and download the developer YAML files in the directory you want to store them. This example uses `~/cht-4-app-developer` as the directory:

```shell 
mkdir  ~/cht-4-app-developer && cd ~/cht-4-app-developer
curl -s -o docker-compose.yml https://raw.githubusercontent.com/medic/cht-upgrade-service/main/docker-compose.yml
curl -s -o cht-core.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic%3Amedic%3Amaster/docker-compose/cht-core.yml
curl -s -o cht-couchdb.yml https://staging.dev.medicmobile.org/_couch/builds_4/medic%3Amedic%3Amaster/docker-compose/cht-couchdb.yml
```

You should now have 3 compose files which we can check with `ls`:

```shell
ls
cht-core.yml  cht-couchdb.yml  docker-compose.yml
```

To start the first developer CHT instance, run `docker-compose`, prepending the needed environment variables:

```shell script
CHT_COMPOSE_PROJECT_NAME=app-devl COUCHDB_SECRET=foo DOCKER_CONFIG_PATH=${PWD} COUCHDB_DATA=${PWD}/couchd CHT_COMPOSE_PATH=${PWD} COUCHDB_USER=medic COUCHDB_PASSWORD=password docker-compose up
```

> This may take some minutes to fully start depending on the speed of the Internet connection and speed of the bare-metal host. This is because docker needs to download all the storage layers for all the containers and the CHT needs to run the first run set up. After downloads and setup has completed, the CHT should be accessible on [https://localhost](https://localhost). You can log in with username `medic` and password `password`.

Reviewer: ```suggestion
This may take some minutes to fully start depending on the speed of the internet connection and speed of the host. This is because docker needs to download all the storage layers for all the containers and the CHT needs to run the first run set up. After downloads and setup has completed, the CHT should be accessible on [https://localhost](https://localhost). You can log in with username `medic` and password `password`.
```

Using the above `the_second` sample project, we can create a file `~/the_second/.env` with this contents:

```shell
NGINX_HTTP_PORT=8081 
NGINX_HTTPS_PORT=8444 
CHT_COMPOSE_PROJECT_NAME=second 
COUCHDB_SECRET=foo 
DOCKER_CONFIG_PATH=./
COUCHDB_DATA=./couchd 
CHT_COMPOSE_PATH=./
COUCHDB_USER=medic 
COUCHDB_PASSWORD=password
```

Now it's easy to boot this environment:

```shell
cd ~/the_second
docker-compose up
```

## Switching & concurrent projects

The easiest way to switch between projects is to stop the first set of containers and start the second set. Cancel the first project running in the foreground with `ctrl + c` and `stop` all the project's services:

```shell
docker stop second_api_1 second_cht-upgrade-service_1 second_couchdb_1 second_haproxy_1 second_healthcheck_1 second_nginx_1 second_sentinel_1
```

Alternately, you can stop ALL containers (even non-CHT ones!) with `docker kill $(docker ps -q)`. Then start the other CHT project using either the `.env` file or use the explicit command with ports and other environment variables as shown above.

> To run projects concurrently, instead of cancelling and `stop`ing the first one, open a second terminal and start the second project.  Be careful you don't have port conflicts!

Reviewer: ```suggestion
To run projects concurrently open a second terminal and start the second project so you don't have to cancel and `stop` the first project.  Remember to avoid port conflicts!
```

The `cht-docker-compose.sh` scripts downloads 3 compose files and builds an `.env` file used above. This greatly eases starting your first CHT instance with a simple text based GUI:

![The cht-docker-compose.sh script showing the URL and version of the CHT instance as well as number of containers launched, global container count, medic images downloaded count and OS load average. Finally a "Successfully started my_first_project" message is shown and denotes the login is "medic" and the password is "password".](cht-docker-helper.png)

This script brings a lot of benefits with it:
* You only have to download one bash script
* It has been tested to work on Windows, macOS and Linux
* All compose files and images will be downloaded automatically for you
* All networks, storage volumes and containers will be created 
* A valid TLS certificate will be installed, allowing you to easily test on with CHT Android natively on a mobile device

Currently, the CHT Docker Helper for 4.0 is in beta, and comes with some caveats:
* it is not yet available in `master`
* it does not account for port conflicts, so you need to manually resolve them yourself by editing the `*.env` files

> * upgrades are not yet support

Reviewer: ```suggestion
* upgrades are not yet supported
```

---
title: "App Developer Hosting in CHT 4.x"
linkTitle: "App Developer Hosting"
weight: 40
aliases:
- /apps/guides/hosting/app-developer
description: >
  Hosting the CHT when developing apps
---

{{% alert title="Note" %}} This guide assumes you are a CHT app developer wanting to either run concurrent instances of the CHT, or easily be able to switch between different instances without losing any data while doing so. To do development on the CHT Core Framework itself, see the [development guide]({{< relref "apps/guides/hosting/core-developer.md" >}}).

> To deploy the CHT 3.x in production, see either [AWS hosting]({{< relref "apps/guides/hosting/3.x/self-hosting.md" >}}) or [Self hosting]({{< relref "apps/guides/hosting/3.x/ec2-setup-guide.md" >}}). 4.x production hosting guides are coming soon!{{% /alert %}}

Reviewer: Switch the URLs over...

```suggestion
To deploy the CHT 3.x in production, see either [AWS hosting]({{< relref "apps/guides/hosting/3.x/ec2-setup-guide.md" >}}) or [Self hosting]({{< relref "apps/guides/hosting/3.x/self-hosting.md" >}}). 4.x production hosting guides are coming soon!{{% /alert %}}
```
